eth-rpc/client.nim

```diff
@@ -1,43 +1,46 @@
-import asyncnet, asyncdispatch, tables, json, macros
+import tables, json, macros
+import asyncdispatch2
 import jsonmarshal
 
 type
   RpcClient* = ref object
-    socket: AsyncSocket
+    transp: StreamTransport
     awaiting: Table[string, Future[Response]]
-    address: string
-    port: Port
+    address: TransportAddress
     nextId: int64
   Response* = tuple[error: bool, result: JsonNode]
 
-
 proc newRpcClient*(): RpcClient =
-  ## Creates a new ``RpcClient`` instance. 
-  RpcClient(
-    socket: newAsyncSocket(),
-    awaiting: initTable[string, Future[Response]](),
-    nextId: 1
-  )
+  ## Creates a new ``RpcClient`` instance.
+  result = RpcClient(awaiting: initTable[string, Future[Response]](), nextId: 1)
 
-proc call*(self: RpcClient, name: string, params: JsonNode): Future[Response] {.async.} =
+proc call*(self: RpcClient, name: string,
+           params: JsonNode): Future[Response] {.async.} =
   ## Remotely calls the specified RPC method.
   let id = $self.nextId
   self.nextId.inc
-  let msg = $ %{"jsonrpc": %"2.0", "method": %name, "params": params, "id": %id} & "\c\l"
-  await self.socket.send(msg)
+  let msg = $ %{"jsonrpc": %"2.0", "method": %name, "params": params,
+                "id": %id} & "\c\l"
+  let res = await self.transp.write(msg)
+  assert(res == len(msg))
```

Comment:
This looks like a technical debt. Can't we just use `writeExact` or smth?

Reply:
You never can know that you have send exactly requested amount of bytes, e.g. remote peer disconnects while we are sending message to it. `write()` procedure will of course try to send all available data, but as explained above, in some situations `write()` could not be completed successfully.

Reply:
Transport failure will fail the future and thus raise in the `await`. Is there any chance when `res != len(msg)`. If not, maybe `write` should return `Future[void]` instead? Or additionally be renamed to `writeExact`?

Reply:
`write()` returns Future[int] for a reason, so application can decide is it critical error, or not. When Future[void] is returned application could not know what happens and need to parse OSError to understand is it critical or not. Also there like 50% of chance not to send whole data, everything depends on network connection (remember we are doing this for mobiles, which can lost connection at any moment).

So this `assert` was made as `TODO:`, e.g. what RPC server must do if it not able to send whole data, maybe it needs to disconnect such client, or maybe it just needs to ignore it.

Reply:
`asserts` are not to-do items in general, so we better place an explicit TODO comment on the line above.